segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -584,49 +585,64 @@ public void mergeSegmentsBatch(WireCommands.MergeSegmentsBatch mergeSegments) {
         Futures.allOfWithResults(sources.stream().map(source -> Futures.handleCompose(getSegmentEventCount(source)
-                .thenCompose(eventCount -> {
-                    AttributeUpdateCollection attributeUpdates = new AttributeUpdateCollection();
-                    attributeUpdates.add(new AttributeUpdate(EVENT_COUNT, AttributeUpdateType.Accumulate, eventCount));
-                    return segmentStore.mergeStreamSegment(mergeSegments.getTargetSegmentId(), source, attributeUpdates, TIMEOUT);
-                }), (r, e) -> {
-            if (e != null) {
-                Throwable unwrap = Exceptions.unwrap(e);
-                if (unwrap instanceof StreamSegmentMergedException) {
-                   log.info(mergeSegments.getRequestId(), "Stream segment already merged '{}'.", source);
-                   return segmentStore.getStreamSegmentInfo(mergeSegments.getTargetSegmentId(), TIMEOUT).thenApply(SegmentProperties::getLength);
-                }
-                if (unwrap instanceof StreamSegmentNotExistsException) {
-                    StreamSegmentNotExistsException ex = (StreamSegmentNotExistsException) unwrap;
-                    if (ex.getStreamSegmentName().equals(source)) {
+            .thenCompose(eventCount -> {
+                AttributeUpdateCollection attributeUpdates = new AttributeUpdateCollection();
+                attributeUpdates.add(new AttributeUpdate(EVENT_COUNT, AttributeUpdateType.Accumulate, eventCount));
+                return segmentStore.mergeStreamSegment(mergeSegments.getTargetSegmentId(), source, attributeUpdates, TIMEOUT);
+            }), (r, e) -> {
+                if (e != null) {
+                    Throwable unwrap = Exceptions.unwrap(e);
+                    if (unwrap instanceof StreamSegmentMergedException) {
                         log.info(mergeSegments.getRequestId(), "Stream segment already merged '{}'.", source);
                         return segmentStore.getStreamSegmentInfo(mergeSegments.getTargetSegmentId(), TIMEOUT).thenApply(SegmentProperties::getLength);
                     }
+                    if (unwrap instanceof StreamSegmentNotExistsException) {
+                        StreamSegmentNotExistsException ex = (StreamSegmentNotExistsException) unwrap;
+                        if (ex.getStreamSegmentName().equals(source)) {
+                            log.info(mergeSegments.getRequestId(), "Stream segment already merged '{}'.", source);
+                            return segmentStore.getStreamSegmentInfo(mergeSegments.getTargetSegmentId(), TIMEOUT).thenApply(SegmentProperties::getLength);
+                        }
+                    }
+                    throw new CompletionException(e);
+                } else {
+                    recordStatForTransaction(r, mergeSegments.getTargetSegmentId());
+                    return appendOnIndexSegment(mergeSegments.getTargetSegmentId()).thenApply(v -> r.getTargetSegmentLength());
```

Comment:
Can we use thenCompose() instead of thenApply() as we are not using result of appendOnIndexSegment?

Reply:
If we use thenCompose() then we need to explicitly return a CompletableFuture having **r.getTargetSegmentLength()**. To avoid that, used thenApply here.